src/controllers/krate/publish.rs

```diff
@@ -221,14 +221,34 @@ pub async fn publish(app: AppState, req: BytesRequest) -> AppResult<Json<GoodCra
             return Err(cargo_err("expected at most 5 categories per crate"));
         }
 
+        let max_features = existing_crate
+            .and_then(|c| c.max_features.map(|mf| mf as usize))
+            .unwrap_or(app.config.max_features);
+
         let features = tarball_info.manifest.features.unwrap_or_default();
+        let num_features = features.len();
+        if num_features > max_features {
+            return Err(cargo_err(&format!(
+                "Maximum number of features was reached (uploaded: {}, limit: {})",
+                num_features, max_features
+            )));
+        }
+
         for (key, values) in features.iter() {
             if !Crate::valid_feature_name(key) {
                 return Err(cargo_err(&format!(
                     "\"{key}\" is an invalid feature name (feature names must contain only letters, numbers, '-', '+', or '_')"
                 )));
             }
 
+            let num_features = values.len();
```

Comment:
I think this only happens if we use a bunch of optional deps. Because if you have more than 200 real features then it would fail in the first check. So could you please add some comments to explain it? Please correct me if I misunderstand it.

Reply:
if you have a dependency on another crate that has 5000 features, then in your crate you can have a feature that enables e.g. 1000 of the features of the dependency. if you do this in an application then that's fine, but if you do it in a library and publish that to crates.io we run into similar issues to the issue report linked above.

Reply:
Oh, I see. Do you mean we depend on one crate that doesn't follow this limitation(Released before this PR)? Then in our index file, we have to include all the features of that dependency.

But shouldn't we check the features field from the dependency instead of the features from this package?

Reply:
> Do you mean we depend on one crate that doesn't follow this limitation(Released before this PR)?

not necessarily. Imagine you are writing library `x` and you depend on libraries `y` and `z`. In your `x/Cargo.toml` you could have:

```toml
[features]
all_the_things = [
  "y/feature1",
  "y/feature2",
  "y/feature3",
  # ...
  "z/feature1",
  "z/feature2",
  "z/feature3",
  # ...
]
```

so while `y` and `z` each have less than 300 features, their combination can have more than that.